pilot/pkg/networking/plugin/authn/authentication.go

```diff
@@ -246,42 +248,47 @@ func BuildAuthNFilter(policy *authn.Policy) *http_conn.HttpFilter {
 
 // buildSidecarListenerTLSContext adds TLS to the listener if the policy requires one.
 func buildSidecarListenerTLSContext(authenticationPolicy *authn.Policy, match *ldsv2.FilterChainMatch) *auth.DownstreamTlsContext {
-	if match != nil && match.TransportProtocol == EnvoyRawBufferMatch {
-		return nil
-	}
-	if requireTLS, mTLSParams := RequireTLS(authenticationPolicy); requireTLS {
-		return &auth.DownstreamTlsContext{
-			CommonTlsContext: &auth.CommonTlsContext{
-				TlsCertificates: []*auth.TlsCertificate{
-					{
-						CertificateChain: &core.DataSource{
-							Specifier: &core.DataSource_Filename{
-								Filename: model.AuthCertsPath + model.CertChainFilename,
-							},
+	tlsContext := &auth.DownstreamTlsContext{
+		CommonTlsContext: &auth.CommonTlsContext{
+			TlsCertificates: []*auth.TlsCertificate{
+				{
+					CertificateChain: &core.DataSource{
+						Specifier: &core.DataSource_Filename{
+							Filename: model.AuthCertsPath + model.CertChainFilename,
 						},
-						PrivateKey: &core.DataSource{
-							Specifier: &core.DataSource_Filename{
-								Filename: model.AuthCertsPath + model.KeyFilename,
-							},
+					},
+					PrivateKey: &core.DataSource{
+						Specifier: &core.DataSource_Filename{
+							Filename: model.AuthCertsPath + model.KeyFilename,
 						},
 					},
 				},
-				ValidationContextType: &auth.CommonTlsContext_ValidationContext{
-					ValidationContext: &auth.CertificateValidationContext{
-						TrustedCa: &core.DataSource{
-							Specifier: &core.DataSource_Filename{
-								Filename: model.AuthCertsPath + model.RootCertFilename,
-							},
+			},
+			ValidationContextType: &auth.CommonTlsContext_ValidationContext{
+				ValidationContext: &auth.CertificateValidationContext{
+					TrustedCa: &core.DataSource{
+						Specifier: &core.DataSource_Filename{
+							Filename: model.AuthCertsPath + model.RootCertFilename,
 						},
 					},
 				},
-				// Same as ListenersALPNProtocols defined in listener. Need to move that constant else where in order to share.
-				AlpnProtocols: []string{"h2", "http/1.1"},
-			},
-			RequireClientCertificate: &types.BoolValue{
-				Value: !(mTLSParams != nil && mTLSParams.AllowTls),
 			},
+			// Same as ListenersALPNProtocols defined in listener. Need to move that constant else where in order to share.
+			AlpnProtocols: []string{"h2", "http/1.1"},
+		},
+		RequireClientCertificate: &types.BoolValue{
+			Value: true,
+		},
+	}
+	// We already decide the purpose of the filter chain, ALPN istio traffic or legacy traffic.
+	if match != nil {
```

Comment:
Keep these conditions at the top please: there's no reason to build the whole structure if we'll just throw it away part of the time.

```go
if match == nil || !reflect.DeepEqual(match.ApplicationProtocols, common.ALPNInMesh) {
    return nil
}
if requireTLS, _ := RequireTLS(authenticationPolicy); !requireTLS {
    return nil
}

return &auth.DownstreamTlsContext{
 // ...
}
```

Reply:
These conditions aren't equivalent to what I want to have.

In my implementation, if `match == nil`, we fall backs to check `RequireTLS`, not returning nil directly.

I think it's okay to sacrifice a little redundant work for readability.

Reply:
Ah sorry, then update the condition to do what you want (`if match != nil && !reflect.DeepEqual`).

Either way I think things are more readable if you put the checks first and exit early rather than doing work and throwing it away. Big +1 on removing the nesting though; just rewrite the conditions to exit early and keep things un-nested as they are in this PR.